ibm/resource_ibm_resource_key.go

```diff
@@ -184,28 +263,32 @@ func resourceIBMResourceKeyUpdate(d *schema.ResourceData, meta interface{}) erro
 }
 
 func resourceIBMResourceKeyRead(d *schema.ResourceData, meta interface{}) error {
-	rsContClient, err := meta.(ClientSession).ResourceControllerAPI()
+	rsContClient, err := meta.(ClientSession).ResourceControllerV2API()
 	if err != nil {
 		return err
 	}
 	resourceKeyID := d.Id()
+	resourceKeyGet := rc.GetResourceKeyOptions{
+		ID: &resourceKeyID,
+	}
 
-	resourceKey, err := rsContClient.ResourceServiceKey().GetKey(resourceKeyID)
+	resourceKey, resp, err := rsContClient.GetResourceKey(&resourceKeyGet)
 	if err != nil {
-		return fmt.Errorf("Error retrieving resource key: %s", err)
+		return fmt.Errorf("Error retrieving resource key: %s with resp code: %s", err, resp)
 	}
-	d.Set("credentials", Flatten(resourceKey.Credentials))
+	d.Set("credentials", resourceKey.Credentials)
 	d.Set("name", resourceKey.Name)
 	d.Set("status", resourceKey.State)
-	if roleCrn, ok := resourceKey.Credentials["iam_role_crn"].(string); ok {
-
+	iamRoleCrn := resourceKey.Credentials.IamRoleCRN
+	if *iamRoleCrn != "" {
+		roleCrn := *iamRoleCrn
 		roleName := roleCrn[strings.LastIndex(roleCrn, ":")+1:]
 
-		if strings.Contains(roleCrn, ":customRole:") {
+		if strings.Contains(roleCrn, ":serviceRole:") {
 			iampapv2Client, err := meta.(ClientSession).IAMPAPAPIV2()
 			if err == nil {
-				resourceCRN := resourceKey.Crn.ServiceName
-				roles, err := iampapv2Client.IAMRoles().ListCustomRoles(resourceKey.AccountID, resourceCRN)
+				resourceCRN := resourceKey.CRN
```

Comment:
Service Name is used to list custom roles..
Earlier CRN was object but here it is a string..so have to obtain service name from the CRN string
```
var resourceCRN string
if instance.CRN != nil {
		serviceName := strings.Split(*instance.CRN, ":")
		if len(serviceName) > 4 {
			resourceCRN =serviceName[4]
		}
	}
```

Reply:
done